cpp/src/aggregation/aggregation.cpp

```diff
@@ -22,142 +22,489 @@
 
 namespace cudf {
 
-std::vector<aggregation::Kind> aggregation::get_simple_aggregations(data_type col_type) const
+namespace detail {
+
+// simple_aggregations_collector ----------------------------------------
+
+std::vector<std::unique_ptr<aggregation>> simple_aggregations_collector::visit(
+  data_type col_type, aggregation const& agg)
+{
+  std::vector<std::unique_ptr<aggregation>> aggs;
+  aggs.push_back(agg.clone());
+  return aggs;
+}
+
+std::vector<std::unique_ptr<aggregation>> simple_aggregations_collector::visit(
+  data_type col_type, sum_aggregation const& agg)
+{
+  return visit(col_type, static_cast<aggregation const&>(agg));
```

Comment:
Is this static_cast necessary for the downcast?

Reply:
Oh, I see, you're doing the cast to select the right overload, otherwise this would just be recursive.